knative-operator/pkg/common/telemetry/telemetry.go

```diff
@@ -0,0 +1,121 @@
+package telemetry
+
+import (
+	"time"
+
+	"github.com/openshift-knative/serverless-operator/knative-operator/pkg/common"
+	"k8s.io/apimachinery/pkg/runtime"
+	kafkasourcev1beta1 "knative.dev/eventing-contrib/kafka/source/pkg/apis/sources/v1beta1"
+	eventingsourcesv1beta1 "knative.dev/eventing/pkg/apis/sources/v1beta1"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var log = common.Log.WithName("telemetry")
+
+type Telemetry struct {
+	name string
+	stop chan struct{}
+	tc   *controller.Controller
+	// Protects from processing order, if true we should install telemetry
+	// if it is false we need to uninstall in the next delete stage.
+	// We start by assuming no telemetry is available.
+	shouldInstallTelemetry bool
+}
+
+func NewTelemetry(name string, mgr manager.Manager, objects []runtime.Object, api client.Client) (*Telemetry, error) {
+	t := &Telemetry{}
+	t.name = name
+	t.stop = make(chan struct{})
+	t.shouldInstallTelemetry = true
+	tc, err := newTelemetryController(name, objects, mgr, t, api)
+	if err != nil {
+		return nil, err
+	}
+	t.tc = tc
+	return t, nil
+}
+
+// TryStartTelemetry setups telemetry per component either Eventing, KnativeKafka or Serving.
+// When called it assumes that the component has status ready.
+func (t *Telemetry) TryStartTelemetry(api client.Client, mgr manager.Manager) error {
+	if t == nil {
+		return nil
+	}
+	if t.shouldInstallTelemetry {
+		log.Info("starting telemetry for:", "component", t.name)
+		errCh := make(chan error, 1)
+		// Initialize metrics before we start the corresponding controller.
+		// There is a tiny window to miss events here, but should be ok for telemetry purposes.
+		t.initializeAndUpdateMetrics(api)
+		// Start our controller in a goroutine so that we do not block.
+		go func() {
+			defer close(errCh)
+			// Block until our controller manager is elected leader. We presume our
+			// entire process will terminate if we lose leadership, so we don'telemetry need
+			// to handle that.
+			<-mgr.Elected()
+			// Start our controller. This will block until it is stopped
+			// or the controller returns a starting error.
+			if err := (*t.tc).Start(t.stop); err != nil {
+				log.Error(err, "cannot start telemetry controller for", "component", t.name)
+				errCh <- err
+			}
+		}()
+		select {
+		case err := <-errCh:
+			t.shouldInstallTelemetry = true
+			return err
+		case <-time.After(1 * time.Second): // assume no error, as controller has blocked
```

Comment:
This will randomly block the reconcile loop for 1 second :thinking:. Should we "just" resort on error logging instead?

Reply:
I am trying to find some way to extract the error. I can remove it.